src/Microsoft.AspNet.Mvc.ModelBinding/Validation/DataAnnotationsClientModelValidatorProvider.cs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding.Validation
+{
+    /// <summary>
+    /// An implementation of <see cref="IClientModelValidatorProvider"/> which providers client validators
+    /// for attributes which derive from <see cref="ValidationAttribute"/>. It also provides
+    /// a validator for types which implement <see cref="IClientModelValidator"/>.
+    /// The logic to support <see cref="IClientModelValidator"/>
+    /// is implemented in <see cref="DataAnnotationsClientModelValidator"/>.
+    /// </summary>
+    public class DataAnnotationsClientModelValidatorProvider : IClientModelValidatorProvider
+    {
+        // A factory for validators based on ValidationAttribute.
+        internal delegate IClientModelValidator
+            DataAnnotationsClientModelValidationFactory(ValidationAttribute attribute);
+
+        // Factories for validation attributes
+        private static readonly DataAnnotationsClientModelValidationFactory _defaultAttributeFactory =
+            (attribute) => new DataAnnotationsClientModelValidator(attribute);
+
+        private readonly Dictionary<Type, DataAnnotationsClientModelValidationFactory> _attributeFactories =
+            BuildAttributeFactoriesDictionary();
+
+        internal Dictionary<Type, DataAnnotationsClientModelValidationFactory> AttributeFactories
+        {
+            get { return _attributeFactories; }
+        }
+
+        /// <inheritdoc />
+        public void GetValidators(ClientValidatorProviderContext context)
+        {
+            foreach (var attribute in context.ValidatorMetadata.OfType<ValidationAttribute>())
+            {
+                DataAnnotationsClientModelValidationFactory factory;
+                if (!_attributeFactories.TryGetValue(attribute.GetType(), out factory))
+                {
+                    factory = _defaultAttributeFactory;
```

Comment:
There is another piece that I wanted to do (as discussed) this should expose the dictionary in some way will be doing this in a followup PR.